queue_consumer.go

```diff
@@ -132,30 +151,48 @@ func (mf *Consumer) Run(ctx context.Context) error {
 	del := NewBatchDeleter(cleanupCtx, cleanupWG, mf.s, mf.DeleteMessageAccumulatorTimeout, mf.DeleteMessageDrainTimeout)
 
 	messages := make(chan job)
-	go mf.receiveMessages(cleanupCtx, cleanupWG, messages, del)
-	var done bool
+	go mf.receiveMessages(cleanupCtx, cleanupWG, mf.shutDown, messages, del)
+	defer func() {
+		close(jobs)
+		wg.Wait()
+		cleanupCancel()
+		cleanupWG.Wait()
+	}()
```

Comment:
It's hard to check the correctness of code when you start passing around WaitGroups and can't easily confirm every `Add()` results in the required number of matching calls to `Done()`. I'm not sure adding even more asynchrony with a timeout is the solution, but I'd feel better seeing that than a deferred function which can potentially deadlock.

Reply:
Agreed.  As we discussed off Github, I will submit another PR to tackle simplifying the WaitGroup usage.